src/util/simplify_expr.cpp

```diff
@@ -235,6 +235,140 @@ bool simplify_exprt::simplify_function_application(exprt &expr)
     expr = from_integer(is_prefix ? 1 : 0, expr.type());
     return false;
   }
+  else if(func_id == ID_cprover_string_endswith_func)
+  {
+    const refined_string_exprt &s1 =
+      to_string_expr(function_app.arguments().at(0));
+
+    const auto s1_data_opt = try_get_string_data_array(s1, ns);
+
+    if(!s1_data_opt)
+    {
+      return true;
+    }
+
+    const array_exprt &s1_data = s1_data_opt->get();
+
+    const refined_string_exprt &s2 =
+      to_string_expr(function_app.arguments().at(1));
+
+    const auto s2_data_opt = try_get_string_data_array(s2, ns);
+
+    if(!s2_data_opt)
+    {
+      return true;
+    }
+
+    const array_exprt &s2_data = s2_data_opt->get();
+
+    bool res = false;
+
+    if(s2_data.operands().size() > s1_data.operands().size())
+    {
+      res = false;
+    }
+    else
+    {
+      res = std::equal(
+        s2_data.operands().rbegin(),
+        s2_data.operands().rend(),
+        s1_data.operands().rbegin());
+    }
+
+    expr = from_integer(res ? 1 : 0, expr.type());
+
+    return false;
+  }
+  else if(func_id == ID_cprover_string_is_empty_func)
+  {
+    const refined_string_exprt &s =
+      to_string_expr(function_app.arguments().at(0));
+
+    const auto s_data_opt = try_get_string_data_array(s, ns);
```

Comment:
I'd expect a null data array to be valid too, especially for a constant empty string (Java strings are counted so don't need null terminators) -- how about just checking the length field instead?

Reply:
Changed to checking the length